tensorflow/python/ops/ragged/row_partition.py

```diff
@@ -746,19 +747,26 @@ def row_splits(self):
     """
     return self._row_splits
 
-  def value_rowids(self):
+  def value_rowids(self, cache=True):
     """Returns the row indices for this row partition.
 
     `value_rowids` specifies the row index fo reach value.  In particular,
     `value_rowids[i]` is the row index for `values[i]`.
 
+    Args:
+      cache: If True, the generated value_rowids tensor will be 
+      stored internally and reused if needed again. Defaults to True.
+
     Returns:
       A 1-D integer `Tensor` with shape `[self.nvals()]`.
       The returned tensor is nonnegative, and is sorted in ascending order.
     """
     if self._value_rowids is not None:
       return self._value_rowids
-    return segment_id_ops.row_splits_to_segment_ids(self._row_splits)
+    value_rowids = segment_id_ops.row_splits_to_segment_ids(self._row_splits)
+    if cache:
+      self._value_rowids = value_rowids
```

Comment:
Thanks for the CL!
I think cache the tensor at python level could hit out of scope error if value_rowids is a graph tensor. For example:

```
class Test:
  def __init__(self, x):
    self.x = x
    self.value = None

  @property
  def row_ids(self):
    if self.value is not None:
      return self.value

      self.value = self.x + 1

    return self.value

@tf.function
def f(x):
  t = Test(x)
  if x > 0:
    res = t.row_ids  # Cache graph tensor t.value in control flow graph
  else:
    res = x

  return res + t.row_ids  # Error: Access graph tensor t.value from another graph

f(tf.constant(1))
```

In this example, the `t.value` we cached was from the control flow graph and we are trying to access it in the return statement in outer f graph. This caused the out of scope error. This is because at the boundary of graph we pass RaggedTensors as values and row_splits. Other encodings like row_ids are not outputs. So trying to access them in other graphs is not allowed.

Is the performance improvement mostly in eager mode? In eager mode, I think it makes sense to cache the value. 
If so, we could do something like

```
if isinstance(value_rowids, ops.EagerTensor):
  self._value_rowids = value_rowids
```

Reply:
I'm not sure we want to treat Eager mode specially here.  That will make this code harder to reason about and maintain.  Plus, performance is more important for graph mode.

Reply:
Thanks!  Confirmed I was able to induce an out-of-scope error with this change using:

```
@tf.function
def ragged_test(x):
    r = tf.RaggedTensor.from_tensor(x)
    if x[0][0] > 0:
        res = r.value_rowids()
    else:
        res = x[0][0]
    return res + r.value_rowids()

ragged_test(tf.constant([[1, 1, 2], [1, 0, 0]], dtype=tf.int64))
```

I did expect that this change would primarily benefit performance in eager mode, but upon testing a simple case with value_rowids() called 1000x for the same RaggedTensor, it seems that speedup is quite similar for both (38s to 0.3s in eager mode, 44s to 1.2s in graph mode).

One side question: is a similar out of scope error possible when creating a RaggedTensor/RowPartition with the from_row_lengths or from_value_rowids methods, which cache those tensors in the constructor?  I didn't encounter any error with a couple of quick experiments, but wanted to confirm.

Reply:
The `row_lengths` cached in `RowPartition.from_row_lengths` is usually less of a problem. Because the are input to the the method. The `row_ids` here is created inside ragged tensor method, which could be called anywhere. If it is called inside a graph, the cached graph tensor `row_ids` will be smuggled out of the graph and cause the https://www.tensorflow.org/guide/function#all_outputs_of_a_tffunction_must_be_return_values.

Could you share a colab for the graph mode test? I was expecting CSE to remove unnecessary conversions. We will need to fix that if it is not the case.

Reply:
If you expect `value_rowids` to be called many times, you could also update the row_parition to have cached `value_rowids`:
```
rt = tf.ragged.constant([[1, 2], [3]]) # Some ragged tensor only with row splits
rt._row_partition = rt._row_partition._with_precomputed_value_rowids()
rt.value_rowids()  # Should use the cached value
```

Reply:
Thanks for clarifying!  I'll do some tests with the row partition functions.

Colab notebook with the test I used for graph vs eager mode is [here](https://colab.research.google.com/drive/1SP_Ul_XqW4omHCgA6-nIdOA08sXclF8H?usp=sharing); with the GPU runtime here I'm seeing that graph mode takes about 30s while eager mode takes about 60s.